networking/src/Ntp/Client.hs

```diff
@@ -7,114 +8,112 @@
+      -- ^ NTP checking interval
+    , ntpTimeDifferenceWarnThreshold :: Microsecond
+      -- ^ Maxium tolerable difference between NTP time and locak time.
     }
 
-hoistNtpClientSettings
-    :: (m () -> n ()) -> NtpClientSettings m -> NtpClientSettings n
-hoistNtpClientSettings f settings =
-    settings { ntpHandler = f . ntpHandler settings }
-
-data NtpClient m = NtpClient
+data NtpClient = NtpClient
     { ncSockets  :: TVar Sockets
+      -- ^ ntp client sockets: ipv4 / ipv6 / both
     , ncState    :: TVar (Maybe [(Microsecond, Microsecond)])
-    , ncSettings :: NtpClientSettings m
+      -- ^ ntp client state, list of received values
+    , ncStatus          :: TVar NtpStatus
+      -- ^ got time callback (margin, time when client sent request)
+    , ncSettings :: NtpClientSettings
+      -- ^ client configuration
     }
 
-mkNtpClient :: MonadIO m => NtpClientSettings m -> Sockets -> m (NtpClient m)
-mkNtpClient ncSettings sock = liftIO $ do
+mkNtpClient :: MonadIO m => NtpClientSettings -> TVar NtpStatus -> Sockets -> m NtpClient
+mkNtpClient ncSettings ncStatus sock = liftIO $ do
     ncSockets <- newTVarIO sock
     ncState  <- newTVarIO Nothing
     return NtpClient{..}
 
-instance Monad m => Default (NtpClientSettings m) where
-    def = NtpClientSettings
-        { ntpServers         = [ "ntp5.stratum2.ru"
-                               , "ntp1.stratum1.ru"
-                               , "clock.isc.org"
-                               ]
-        , ntpHandler         = \_ -> return ()
-        , ntpLogName         = "ntp-cli"
-        , ntpResponseTimeout = 1000000
-        , ntpPollDelay       = 3000000
-        , ntpMeanSelection   = \l -> let len = length l in (sortOn fst l) !! ((len - 1) `div` 2)
-        }
-
 data NoHostResolved = NoHostResolved
     deriving (Show, Typeable)
 
 instance Exception NoHostResolved
 
-type NtpMonad m =
-    ( MonadIO m
-    , MonadBaseControl IO m
-    , MonadMask m
-    , WithLogger m
-    , Mockable Concurrently m
-    , Mockable Async m
-    , Mockable Delay m
-    )
-
-handleCollectedResponses :: NtpMonad m => NtpClient m -> m ()
+usingNtpLogger :: LoggerNameBox IO a -> IO a
+usingNtpLogger = Wlog.usingLoggerName "NtpClient"
+
+logError :: Text -> IO ()
+logError = usingNtpLogger . Wlog.logError
+
+logWarning :: Text -> IO ()
+logWarning = usingNtpLogger . Wlog.logWarning
+
+logInfo :: Text -> IO ()
+logInfo = usingNtpLogger . Wlog.logInfo
+
+logDebug :: Text -> IO ()
+logDebug = usingNtpLogger . Wlog.logDebug
+
+-- |
+-- Handle results.  It is either when all ntp servers responded or when
+-- `ntpResponseTimeout` has passed since the request where send.  If none of the servers responded `ncState`
+handleCollectedResponses :: NtpClient -> IO ()
 handleCollectedResponses cli = do
-    mres <- liftIO $ readTVarIO (ncState cli)
+    mres <- readTVarIO (ncState cli)
     let selection = ntpMeanSelection (ncSettings cli)
-        handler   = ntpHandler (ncSettings cli)
     case mres of
```

Comment:
Since we now have `NtpSyncUnavailable` we can use it when none of the server responded.